README.md
<!-- 
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages). 

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages). 
-->

# open-meteo-dart
A client for [open-meteo](https://open-meteo.com/) API written in dart.
It currently only supports the `forecast` endpoint, but other endpoints are planned in the future.


## Features

- `forecast` API with automatic model selection performed by the server

### Planned features
- other endpoints, see [here](https://open-meteo.com/en/docs):
    - historical weather
    - ensembel models
    - climate change
    - marine forecast
    - air quality
    - geocoding
    - elevation
    - flood
- requesting a specific model atthe `forecast` endpoint

## Getting started

<!-- TODO Add this when the package is available at pub.dev
You can use the command line to add the dependency with the latest stable version:
```shell
$ dart pub add open-meteo-dart
```
-->

<!--
You can manually add the dependency in your `pubspec.yaml`:
```yaml
dependencies:
    open-meteo-dart:
        git: https://github.com/timosturm/open-meteo-dart.git
        ref: stable # TODO inlcude a tag here!
```
-->

You can also use a version from github in `pubspec.yaml`:
```yaml
dependencies:
      open_meteo_dart:
        git: 
          url: https://github.com/timosturm/open-meteo-dart.git
          ref: v0.0.1 # Get version 0.0.1 of the API.
```


## Usage

<!-- TODO Automatically include the example from example/short_example.dart -->
```dart
import 'package:open_meteo_dart/open_meteo_dart.dart';

void main() async {
  OpenMeteoApi api =
      OpenMeteoApi(logging: true); // Print the response to the console.

  ForecastResponseV1 responseV1 = await api.forecastV1(
    latitude: 52.52,
    longitude: 13.41,
    hourlyParameters:
        HourlyParameter.allRequestParameters, 
    dailyParameters: DailyParameter.allRequestParameters,
    currentWeather: true,
    forecastDays: 2,
    pastDays: 2,
  );

  print("Current temperature: ${responseV1.currentWeather?[CurrentWeatherParameter.temperature]}");
  // Current temperature: 24.3

  print("Requested days: ${responseV1.daily?[DailyParameter.time]}");
  // Requested days: [2023-05-29, 2023-05-30, 2023-05-31, 2023-06-01]
  print("Daily apparent max temperature: ${responseV1.daily?[DailyParameter.apparentTemperatureMax]}");
  // Daily apparent max temperature: [19.8, 20.5, 22.6, 20.4]
}
```

A more ellaborate example can be found [here](example/example.dart).

## Additional information

More information about open-meteo can be found at the [official site](https://open-meteo.com/), including information about the weather variables.

For now, I consider this package usable. If you require a certain API that is not yet implemented, please fell free to file an issue or (even better) submit a pull request.

For code / implementation style take a look at the implemented forecast API. I prefer to have all request and response parameters hard coded as an `enum` to allow easier use via code-completion and type-safety. E.g., a user will know what parameters are allowed by the forecast API for daily forecasts by looking at the enum `DailyParameter`.


CHANGELOG.md
## 0.0.1

- Initial version.
- Forecast v1 api.

test/open_meteo_dart_test.dart
import 'package:open_meteo_dart/open_meteo_dart.dart';
import 'package:test/test.dart';

void main() {
  group('A group of tests', () {
    final OpenMeteoApi api = OpenMeteoApi(logging: true);

    // setUp(() {
    //   // Additional setup goes here.
    // });

    // TODO More elaborate tests should be added.
    test('Test forecast api, all paramters', () {
      expectLater(
          api.forecastV1(
            latitude: 52.52,
            longitude: 13.41,
            hourlyParameters: HourlyParameter.allRequestParameters,
            dailyParameters: DailyParameter.allRequestParameters,
            currentWeather: true,
            forecastDays: 2,
            pastDays: 2,
          ),
          isA<ForecastResponseV1>());
    });
  });
}


lib/src/api.dart
import 'package:dio/dio.dart';
import 'package:open_meteo_dart/src/request_parameters/parameters_forecast_v1.dart';

import 'models/response_forecast.dart';

class OpenMeteoApi {
  final String baseURL;
  final String endpointForecastV1;
  final bool logging;

  late final Dio dio;

  OpenMeteoApi({
    Dio? customDio,
    this.baseURL = "https://api.open-meteo.com",
    this.endpointForecastV1 = "/v1/forecast",
    this.logging = true,
  }) {
    if (customDio == null) {
      dio = Dio();

      if (logging) {
        dio.interceptors.add(LogInterceptor(responseBody: true));
      }
    } else {
      dio = customDio;
    }

    // TODO Add some custom error handling and wrap the 400 reponse!
  }

  /// Calls the forecast v1 endpoint.
  ///
  /// TODO Add fine grained support for specific models.
  ///
  /// Params:
  /// - latitude: Geographical WGS84 coordinate of the location.
  /// - longitude: Geographical WGS84 coordinate of the location.
  /// - elevation: The elevation used for statistical downscaling.
  ///              Per default, a [90 meter digital elevation model](https://openmeteo.substack.com/p/improving-weather-forecasts-with) is used.
  ///              You can manually set the elevation to correctly match mountain peaks.
  ///              If elevation is not specified, downscaling will be disabled and the
  ///              API uses the average grid-cell height.
  /// - hourly: A list of weather variables which should be returned.
  /// - daily: A list of daily weather variable aggregations which should be returned.
  /// - currentWeather: Weather to include current weather conditions.
  /// - temperatureUnit: [TemperatureUnit] to use.
  /// - windspeedUnit: [WindspeedUnit] to use.
  /// - precipitationUnit:[PrecipitationUnit] to use.
  /// - pastDays: Include data aboute past days in response, valid values between 0 and 92 days.
  /// - forecastDays: Include data about this many days into the future, for [forecastDays] set to 1 returns the current day.
  /// - startDate: The [DateTime] interval to get weather data.
  /// - endDate: The [DateTime] interval to get weather data.
  /// - TODO models auto	String array 	No 	auto 	Manually select one or more weather models. Per default, the best suitable weather models will be combined.
  /// - cellSelection: Set a preference how grid-cells are selected. The default [CellSelection.land] finds a suitable grid-cell on land with
  ///                  [similar elevation to the requested coordinates using a 90-meter digital elevation model](https://openmeteo.substack.com/p/improving-weather-forecasts-with).
  ///                  [CellSelection.sea] prefers grid-cells on sea. [CellSelection.nearest] selects the nearest possible grid-cell.
  Future<ForecastResponseV1> forecastV1({
    required double latitude,
    required double longitude,
    double? elevation, // TODO Use 90 or null as the default?
    List<DailyParameter>? dailyParameters,
    List<HourlyParameter>? hourlyParameters,
    bool currentWeather = false,
    TemperatureUnit temperatureUnit = TemperatureUnit.celsius,
    WindspeedUnit windspeedUnit = WindspeedUnit.kmh,
    PrecipitationUnit precipitationUnit = PrecipitationUnit.mm,
    int? pastDays,
    int? forecastDays,
    DateTime? startDate, // TODO Use datetime
    DateTime? endDate, // TODO Use datetime
    // String models = "auto", // TODO Make an enum
    CellSelection cellSelection = CellSelection.land,
  }) async {
    if (forecastDays != null) {
      assert(0 <= forecastDays && forecastDays <= 16);
    }

    if (pastDays != null) {
      assert(0 <= pastDays && pastDays <= 92);
    }

    // Either start_date and end_date must be given or one of forecast_days or past_days.
    assert((startDate != null && endDate != null) ||
        (forecastDays != null || pastDays != null));

    // TODO past_days and forecast_days are mutually exclusive with start_date and end_date.
    if (startDate != null || endDate != null) {
      // Both must be given at the same time.
      assert(startDate != null && endDate != null);
      assert(pastDays == null && forecastDays == null);
    }

    // time must not be part of the request!
    assert(!(hourlyParameters?.contains(HourlyParameter.time) ?? false));
    assert(!(dailyParameters?.contains(DailyParameter.time) ?? false));

    final Map<String, dynamic> queryParameters = {
      "latitude": latitude,
      "longitude": longitude,
      "elevation": elevation ?? "nan",
      "daily":
          dailyParameters?.expand((element) => [element.variable]).join(","),
      "hourly": hourlyParameters?.expand((e) => [e.variable]).join(","),
      "current_weather": currentWeather,
      "temperature_unit": temperatureUnit.name,
      "windspeed_unit": windspeedUnit.name,
      "precipitation_unit": precipitationUnit.name,
      "timeformat":
          "iso8601", // Hardcoded value because we work with [DateTime] anyways.
      "timezone":
          "UTC", // FIXME The timezone stuff must be fixed, this will make the api return UTC timezone which must be converted back to local!
      "past_days": pastDays,
      "forecast_days": forecastDays,
      "start_date": startDate
          ?.toUtc()
          .toIso8601String()
          .substring(0, 10), // Only retain the yyyy-mm-dd part.
      "end_date": endDate
          ?.toUtc()
          .toIso8601String()
          .substring(0, 10), // Only retain the yyyy-mm-dd part.
      // "models": models,
      "cell_selection": cellSelection.name,
    };

    final response = await dio.get(
      "$baseURL$endpointForecastV1",
      queryParameters: queryParameters,
    );

    return ForecastResponseV1.fromJson(response.data);
  }
}


example/short_example.dart
import 'package:open_meteo_dart/open_meteo_dart.dart';

void main() async {
  OpenMeteoApi api =
      OpenMeteoApi(logging: true); // Print the response to the console.

  ForecastResponseV1 responseV1 = await api.forecastV1(
    latitude: 52.52,
    longitude: 13.41,
    hourlyParameters:
        HourlyParameter.allRequestParameters,
    dailyParameters: DailyParameter.allRequestParameters,
    currentWeather: true,
    forecastDays: 2,
    pastDays: 2,
  );

  print(
      "Current temperature: ${responseV1.currentWeather?[CurrentWeatherParameter.temperature]}");
  // Current temperature: 24.3

  print("Requested days: ${responseV1.daily?[DailyParameter.time]}");
  // Requested days: [2023-05-29, 2023-05-30, 2023-05-31, 2023-06-01]
  print(
      "Daily apparent max temperature: ${responseV1.daily?[DailyParameter.apparentTemperatureMax]}");
  // Daily apparent max temperature: [19.8, 20.5, 22.6, 20.4]
}


lib/src/models/response_forecast.dart
import 'package:json_annotation/json_annotation.dart';

import 'package:open_meteo_dart/src/request_parameters/parameters_forecast_v1.dart';

part 'response_forecast.g.dart';

/// Object representation of the JSON response from the /v1/forecast endpoint.
@JsonSerializable()
class ForecastResponseV1 {
  /// WGS84 of the center of the weather grid-cell which was used to generate this forecast. This coordinate might be a few kilometers away from the requested coordinate.
  final double latitude;

  /// WGS84 of the center of the weather grid-cell which was used to generate this forecast. This coordinate might be a few kilometers away from the requested coordinate.
  final double longitude;

  /// The elevation from a 90 meter digital elevation model. This effects which grid-cell is selected (see parameter cell_selection). Statistical downscaling is used to adapt weather conditions for this elevation. This elevation can also be controlled with the query parameter elevation. If &elevation=nan is specified, all downscaling is disabled and the averge grid-cell elevation is used.
  final double elevation;

  /// Generation time of the weather forecast in milliseconds. This is mainly used for performance monitoring and improvements.
  @JsonKey(name: "generationtime_ms")
  final double generationTimeMs;

  /// Applied timezone offset from the &timezone= parameter.
  @JsonKey(name: "utc_offset_seconds")
  final int utcOffsetSeconds;

  /// Timezone identifier (e.g. Europe/Berlin)
  final String timezone;

  /// Timezone abbreviation (e.g. CEST)
  @JsonKey(name: "timezone_abbreviation")
  final String timezoneAbbreviation;

  /// Object 	For each selected weather variable, data will be returned as a floating point array. Additionally a time array will be returned with ISO8601 timestamps.
  final Map<HourlyParameter, List>? hourly;

  ///	For each selected weather variable, the unit will be listed here.
  @JsonKey(name: "hourly_units")
  final Map<HourlyParameter, String>? hourlyUnits;

  ///	Object 	For each selected daily weather variable, data will be returned as a floating point array. Additionally a time array will be returned with ISO8601 timestamps.
  final Map<DailyParameter, List>? daily;

  ///	For each selected daily weather variable, the unit will be listed here.
  @JsonKey(name: "daily_units")
  final Map<DailyParameter, String>? dailyUnits;

  ///	Object 	Current weather conditions with the attributes: time, temperature, windspeed, winddirection and weathercode
  @JsonKey(name: "current_weather")
  final Map<CurrentWeatherParameter, dynamic>? currentWeather;

  const ForecastResponseV1(
    this.latitude,
    this.longitude,
    this.elevation,
    this.generationTimeMs,
    this.utcOffsetSeconds,
    this.timezone,
    this.timezoneAbbreviation,
    this.hourly,
    this.hourlyUnits,
    this.daily,
    this.dailyUnits,
    this.currentWeather,
  );

  /// Connect the generated [_$PersonFromJson] function to the `fromJson`
  /// factory.
  factory ForecastResponseV1.fromJson(Map<String, dynamic> json) =>
      _$ForecastResponseV1FromJson(json);

  /// Connect the generated [_$PersonToJson] function to the `toJson` method.
  Map<String, dynamic> toJson() => _$ForecastResponseV1ToJson(this);

  @override
  String toString() {
    return 'ForecastResponse(latitude: $latitude, longitude: $longitude, elevation: $elevation, generationTimeMs: $generationTimeMs, utcOffsetSeconds: $utcOffsetSeconds, timezone: $timezone, timezoneAbbreviation: $timezoneAbbreviation, hourly: $hourly, hourlyUnits: $hourlyUnits, daily: $daily, dailyUnits: $dailyUnits, currentWeather: $currentWeather)';
  }
}


lib/src/models/response_forecast.g.dart
// GENERATED CODE - DO NOT MODIFY BY HAND

part of 'response_forecast.dart';

// **************************************************************************
// JsonSerializableGenerator
// **************************************************************************

ForecastResponseV1 _$ForecastResponseV1FromJson(Map<String, dynamic> json) =>
    ForecastResponseV1(
      (json['latitude'] as num).toDouble(),
      (json['longitude'] as num).toDouble(),
      (json['elevation'] as num).toDouble(),
      (json['generationtime_ms'] as num).toDouble(),
      json['utc_offset_seconds'] as int,
      json['timezone'] as String,
      json['timezone_abbreviation'] as String,
      (json['hourly'] as Map<String, dynamic>?)?.map(
        (k, e) => MapEntry(
            $enumDecode(_$HourlyParameterEnumMap, k), e as List<dynamic>),
      ),
      (json['hourly_units'] as Map<String, dynamic>?)?.map(
        (k, e) =>
            MapEntry($enumDecode(_$HourlyParameterEnumMap, k), e as String),
      ),
      (json['daily'] as Map<String, dynamic>?)?.map(
        (k, e) => MapEntry(
            $enumDecode(_$DailyParameterEnumMap, k), e as List<dynamic>),
      ),
      (json['daily_units'] as Map<String, dynamic>?)?.map(
        (k, e) =>
            MapEntry($enumDecode(_$DailyParameterEnumMap, k), e as String),
      ),
      (json['current_weather'] as Map<String, dynamic>?)?.map(
        (k, e) => MapEntry($enumDecode(_$CurrentWeatherParameterEnumMap, k), e),
      ),
    );

Map<String, dynamic> _$ForecastResponseV1ToJson(ForecastResponseV1 instance) =>
    <String, dynamic>{
      'latitude': instance.latitude,
      'longitude': instance.longitude,
      'elevation': instance.elevation,
      'generationtime_ms': instance.generationTimeMs,
      'utc_offset_seconds': instance.utcOffsetSeconds,
      'timezone': instance.timezone,
      'timezone_abbreviation': instance.timezoneAbbreviation,
      'hourly': instance.hourly
          ?.map((k, e) => MapEntry(_$HourlyParameterEnumMap[k]!, e)),
      'hourly_units': instance.hourlyUnits
          ?.map((k, e) => MapEntry(_$HourlyParameterEnumMap[k]!, e)),
      'daily': instance.daily
          ?.map((k, e) => MapEntry(_$DailyParameterEnumMap[k]!, e)),
      'daily_units': instance.dailyUnits
          ?.map((k, e) => MapEntry(_$DailyParameterEnumMap[k]!, e)),
      'current_weather': instance.currentWeather
          ?.map((k, e) => MapEntry(_$CurrentWeatherParameterEnumMap[k]!, e)),
    };

const _$HourlyParameterEnumMap = {
  HourlyParameter.time: 'time',
  HourlyParameter.temperature2m: 'temperature_2m',
  HourlyParameter.relativehumidity2m: 'relativehumidity_2m',
  HourlyParameter.dewpoint2m: 'dewpoint_2m',
  HourlyParameter.apparentTemperature: 'apparent_temperature',
  HourlyParameter.pressureMsl: 'pressure_msl',
  HourlyParameter.surfacePressure: 'surface_pressure',
  HourlyParameter.cloudcover: 'cloudcover',
  HourlyParameter.cloudcoverLow: 'cloudcover_low',
  HourlyParameter.cloudcoverMid: 'cloudcover_mid',
  HourlyParameter.cloudcoverHigh: 'cloudcover_high',
  HourlyParameter.windspeed10m: 'windspeed_10m',
  HourlyParameter.windspeed80m: 'windspeed_80m',
  HourlyParameter.windspeed120m: 'windspeed_120m',
  HourlyParameter.windspeed180m: 'windspeed_180m',
  HourlyParameter.winddirection10m: 'winddirection_10m',
  HourlyParameter.winddirection80m: 'winddirection_80m',
  HourlyParameter.winddirection120m: 'winddirection_120m',
  HourlyParameter.winddirection180m: 'winddirection_180m',
  HourlyParameter.windgusts10m: 'windgusts_10m',
  HourlyParameter.shortwaveRadiation: 'shortwave_radiation',
  HourlyParameter.directRadiation: 'direct_radiation',
  HourlyParameter.directNormalIrradiance: 'direct_normal_irradiance',
  HourlyParameter.diffuseRadiation: 'diffuse_radiation',
  HourlyParameter.vaporPressureDeficit: 'vapor_pressure_deficit',
  HourlyParameter.cape: 'cape',
  HourlyParameter.evapotranspiration: 'evapotranspiration',
  HourlyParameter.et0FaoEvapotranspiration: 'et0_fao_evapotranspiration',
  HourlyParameter.precipitation: 'precipitation',
  HourlyParameter.snowfall: 'snowfall',
  HourlyParameter.precipitationProbability: 'precipitation_probability',
  HourlyParameter.rain: 'rain',
  HourlyParameter.showers: 'showers',
  HourlyParameter.weathercode: 'weathercode',
  HourlyParameter.snowDepth: 'snow_depth',
  HourlyParameter.freezinglevelHeight: 'freezinglevel_height',
  HourlyParameter.visibility: 'visibility',
  HourlyParameter.soilTemperature0cm: 'soil_temperature_0cm',
  HourlyParameter.soilTemperature6cm: 'soil_temperature_6cm',
  HourlyParameter.soilTemperature18cm: 'soil_temperature_18cm',
  HourlyParameter.soilTemperature54cm: 'soil_temperature_54cm',
  HourlyParameter.soilMoisture0to1cm: 'soil_moisture_0_1cm',
  HourlyParameter.soilMoisture1to3cm: 'soil_moisture_1_3cm',
  HourlyParameter.soilMoisture3to9cm: 'soil_moisture_3_9cm',
  HourlyParameter.soilMoisture9to27cm: 'soil_moisture_9_27cm',
  HourlyParameter.soilMoisture27to81cm: 'soil_moisture_27_81cm',
  HourlyParameter.isDay: 'is_day',
};

const _$DailyParameterEnumMap = {
  DailyParameter.time: 'time',
  DailyParameter.temperature2mMin: 'temperature_2m_min',
  DailyParameter.temperature2mMax: 'temperature_2m_max',
  DailyParameter.apparentTemperatureMin: 'apparent_temperature_min',
  DailyParameter.apparentTemperatureMax: 'apparent_temperature_max',
  DailyParameter.precipitationSum: 'precipitation_sum',
  DailyParameter.rainSum: 'rain_sum',
  DailyParameter.showersSum: 'showers_sum',
  DailyParameter.snowfallSum: 'snowfall_sum',
  DailyParameter.precipitationHours: 'precipitation_hours',
  DailyParameter.precipitationProbabilityMin: 'precipitation_probability_min',
  DailyParameter.precipitationProbabilityMax: 'precipitation_probability_max',
  DailyParameter.precipitationProbabilityMean: 'precipitation_probability_mean',
  DailyParameter.weathercode: 'weathercode',
  DailyParameter.sunrise: 'sunrise',
  DailyParameter.sunset: 'sunset',
  DailyParameter.windspeed10mMax: 'windspeed_10m_max',
  DailyParameter.windgusts10mMax: 'windgusts_10m_max',
  DailyParameter.winddirection10mDominant: 'winddirection_10m_dominant',
  DailyParameter.shortwaveRadiationSum: 'shortwave_radiation_sum',
  DailyParameter.et0FaoEvapotranspiration: 'et0_fao_evapotranspiration',
  DailyParameter.uvIndexMax: 'uv_index_max',
  DailyParameter.uvIndexClearSkyMax: 'uv_index_clear_sky_max',
};

const _$CurrentWeatherParameterEnumMap = {
  CurrentWeatherParameter.time: 'time',
  CurrentWeatherParameter.temperature: 'temperature',
  CurrentWeatherParameter.windspeed: 'windspeed',
  CurrentWeatherParameter.winddirection: 'winddirection',
  CurrentWeatherParameter.weathercode: 'weathercode',
  CurrentWeatherParameter.isDay: 'is_day',
};


example/example.dart
import 'package:open_meteo_dart/open_meteo_dart.dart';

void main() async {
  // Create an instance of the api, customize the used [Dio] instance if you like:
  // ```dart
  // OpenMeteoApi api = OpenMeteoApi(customDio: Dio());
  // ```
  OpenMeteoApi api =
      OpenMeteoApi(logging: true); // Print the response to the console.

  ForecastResponseV1 responseV1 = await api.forecastV1(
    latitude: 52.52,
    longitude: 13.41,
    // hourlyParameters: HourlyParameter.allRequestParameters,  // Request all parameters.
    hourlyParameters: [
      HourlyParameter.temperature2m,
      HourlyParameter.cloudcoverLow,
    ],
    // dailyParameters: DailyParameter.allRequestParameters, // Request all parameters.
    dailyParameters: [
      DailyParameter.apparentTemperatureMax,
      DailyParameter.precipitationSum,
    ],
    currentWeather: true,
    forecastDays: 2,
    pastDays: 2,
  );

  print(
      "Current temperature: ${responseV1.currentWeather?[CurrentWeatherParameter.temperature]}");
  // Current temperature: 24.3

  print("Requested days: ${responseV1.daily?[DailyParameter.time]}");
  // Requested days: [2023-05-29, 2023-05-30, 2023-05-31, 2023-06-01]
  print(
      "Daily apparent max temperature: ${responseV1.daily?[DailyParameter.apparentTemperatureMax]}");
  // Daily apparent max temperature: [19.8, 20.5, 22.6, 20.4]
}
